groups/service.go

```diff
@@ -507,6 +508,46 @@ func (gs *groupService) UpdateGroupSettings(group GoogleGroup) error {
 	return nil
 }
 
+// MigrateMailingListMembers copies members from one Google Group (sourceGroup) to another (destinationGroup)
+// while preserving their existing roles.
+func (as *adminService) MigrateMailingListMembers(sourceGroup, destinationGroup GoogleGroup) error {
+	if *verbose {
+		log.Printf("Copying members from %s to %s", sourceGroup.EmailId, destinationGroup.EmailId)
+	}
+
+	// check if the source group exists
+	sourceGroupMembers, err := as.client.ListMembers(sourceGroup.EmailId)
+	if err != nil {
+		if as.checkForAPIErr404(err) {
+			log.Printf("Source group %q does not exist or has no members\n", sourceGroup.EmailId)
+			return nil
+		}
+		return fmt.Errorf("unable to retrieve members from source group %q: %w", sourceGroup.EmailId, err)
+	}
+
+	// just a hacky way to check if the destination group exists, not using the destination members list
+	_, err = as.client.ListMembers(destinationGroup.EmailId)
+	if err != nil && !as.checkForAPIErr404(err) {
+		return fmt.Errorf("unable to retrieve members from destination group %q: %w", destinationGroup.EmailId, err)
+	}
+
+	// aggregate the errors that occurred and return them together in the end
+	var errs []error
+
+	// iterate through source group members and add/update them in the destination group
+	for _, sourceMember := range sourceGroupMembers {
+		err := as.AddOrUpdateGroupMembers(destinationGroup, sourceMember.Role, []string{sourceMember.Email})
+		if err != nil {
+			logErr := fmt.Errorf("unable to add/update %s to %q as %s: %w", sourceMember.Email, destinationGroup.EmailId, sourceMember.Role, err)
+			log.Printf("%s\n", logErr)
+			errs = append(errs, logErr)
+			continue
+		}
+	}
```

Comment:
Is it possible to replace this loop with a single call like such:
```go
sourceMemberEmailIDs := getEmailIDs(sourceGroupMembers)
err := as.AddOrUpdateGroupMembers(destinationGroup, sourceMember, sourceMemberEmailIDs)
```

Reply:
Two things here:

- In the `AddOrUpdateGroupMembers` function, the second argument is taking the member's `Role`.
   ```
   func (as *adminService) AddOrUpdateGroupMembers(group GoogleGroup, role string, members []string) error {...}
   ```

- So, we're using a loop to process one Mailing List member at a time from `sourceGroupMembers` list.
  We pass their `Role` and `Email` to `AddOrUpdateGroupMembers()` to update the destination group without affecting their existing role.

  ```
  err := as.AddOrUpdateGroupMembers(destinationGroup, sourceMember.Role, []string{sourceMember.Email})
  ```

Please let me know if I misunderstood your suggestion.